# Ex2 - Getting and Knowing your Data

This time we are going to pull data directly from the internet.
Special thanks to: https://github.com/justmarkham for sharing the dataset and materials.

### Step 1. Import the necessary libraries

In [1]:
import numpy as np
import pandas as pd

### Step 2. Import the dataset from this [address](https://raw.githubusercontent.com/justmarkham/DAT8/master/data/chipotle.tsv). 

In [3]:
url = "https://raw.githubusercontent.com/justmarkham/DAT8/master/data/chipotle.tsv"

### Step 3. Assign it to a variable called chipo.

In [4]:
chipo = pd.read_csv(url, sep='\t')

### Step 4. See the first 10 entries

In [5]:
chipo.head(10)

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   
5         3         1                           Chicken Bowl   
6         3         1                          Side of Chips   
7         4         1                          Steak Burrito   
8         4         1                       Steak Soft Tacos   
9         5         1                          Steak Burrito   

                                  choice_description item_price  
0                                                NaN     $2.39   
1                                       [Clementine]     $3.39   
2                                            [Apple]     $3.39   
3                                                NaN     $2.39   
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98   
5  [Fresh Tomato Salsa (Mild), [Rice, Cheese, Sou...    $10.98   
6                                                NaN     $1.69   
7  [Tomatillo Red Chili Salsa, [Fajita Vegetables...    $11.75   
8  [Tomatillo Green Chili Salsa, [Pinto Beans, Ch...     $9.25   
9  [Fresh Tomato Salsa, [Rice, Black Beans, Pinto...     $9.25

### Step 5. What is the number of observations in the dataset?

In [6]:
# Solution 1
len(chipo)


4622

In [8]:
# Solution 2
chipo_shape = chipo.shape
chipo_shape[0]

4622

### Step 6. What is the number of columns in the dataset?

In [9]:
chipo_shape[1]

5

### Step 7. Print the name of all the columns.

In [11]:
for col in chipo.columns:
    print(col)

order_id
quantity
item_name
choice_description
item_price


### Step 8. How is the dataset indexed?

In [12]:
chipo.index

# Because I haven't designated any index upon data (url) import
# i.e. the 'read_csv' method use above, this is simply indexed 
# on the row number in the data itself.

RangeIndex(start=0, stop=4622, step=1)

### Step 9. Which was the most-ordered item? 

In [22]:
chipo_item_counts = chipo.item_name.value_counts()
chipo_item_count_rows = list(chipo_item_counts.iteritems())
chipo_item_count_rows[0][0]

'Chicken Bowl'

_i.e.: 'Chicken Bowl' was the most ordered items, with 726 orders that included this item._

### Step 10. For the most-ordered item, how many items were ordered?

In [58]:
chipo_item_count_rows[0][1]

726

_Step 9, 10 Solutions - First, use `groupby` method on the dataframe:_

In [61]:
g = chipo.groupby('item_name')

In [62]:
g

This is _not_ yet a result like we ordinarily might expect! Instead it is a `DataFrameGroupBy`. How does this look "internally"?

In [63]:
# g is `chipo` dataframe *grouped by*
# `item_name` values
for item_name, item_name_df in g:
    print(item_name, '\n')
    print(item_name_df.head())

6 Pack Soft Drink 

     order_id  quantity          item_name choice_description item_price  \
298       129         1  6 Pack Soft Drink           [Sprite]     $6.49    
341       148         1  6 Pack Soft Drink        [Diet Coke]     $6.49    
357       154         1  6 Pack Soft Drink             [Coke]     $6.49    
388       168         1  6 Pack Soft Drink        [Diet Coke]     $6.49    
417       182         1  6 Pack Soft Drink        [Diet Coke]     $6.49    

    choice_desc_list  
298         [Sprite]  
341      [Diet Coke]  
357           [Coke]  
388      [Diet Coke]  
417      [Diet Coke]  
Barbacoa Bowl 

     order_id  quantity      item_name  \
39         19         1  Barbacoa Bowl   
58         27         1  Barbacoa Bowl   
95         42         1  Barbacoa Bowl   
115        51         1  Barbacoa Bowl   
127        56         1  Barbacoa Bowl   

                                    choice_description item_price  \
39   [Roasted Chili Corn Salsa, [Fajita Vegetab

774  [Tomatillo-Green Chili Salsa (Medium), Pinto B...  
Steak Salad 

      order_id  quantity    item_name  \
664        276         1  Steak Salad   
752        311         1  Steak Salad   
893        369         1  Steak Salad   
3502      1406         1  Steak Salad   

                                     choice_description item_price  \
664   [Tomatillo-Red Chili Salsa (Hot), [Black Beans...     $8.99    
752   [Tomatillo-Red Chili Salsa (Hot), [Black Beans...     $8.99    
893   [Fresh Tomato Salsa (Mild), [Rice, Cheese, Sou...     $8.99    
3502                        [[Lettuce, Fajita Veggies]]     $8.69    

                                       choice_desc_list  
664   [Tomatillo-Red Chili Salsa (Hot), Black Beans,...  
752   [Tomatillo-Red Chili Salsa (Hot), Black Beans,...  
893   [Fresh Tomato Salsa (Mild), Rice, Cheese, Sour...  
3502                          [Lettuce, Fajita Veggies]  
Steak Salad Bowl 

     order_id  quantity         item_name  \
281       123     

So: to each possible `item_name` value that there is within the data set (specifically within the column `chipo['item_name']`), `g` has associated the "sub-dataframe" whose `item_name` value is that one.

Now that we have this `GroupBy` object, we can "sum up" each group which has been made by `item_name`:

In [65]:
g_sum = g.sum()
g_sum

order_id  quantity
item_name                                                
6 Pack Soft Drink                         52322        55
Barbacoa Bowl                             53972        66
Barbacoa Burrito                          74718        91
Barbacoa Crispy Tacos                      5613        12
Barbacoa Salad Bowl                        9708        10
Barbacoa Soft Tacos                       18725        25
Bottled Water                            175944       211
Bowl                                        472         4
Burrito                                    1550         6
Canned Soda                               76396       126
Canned Soft Drink                        304753       351
Carnitas Bowl                             62742        71
Carnitas Burrito                          51652        60
Carnitas Crispy Tacos                      5868         8
Carnitas Salad                             1500         1
Carnitas Salad Bowl                        6693         6
Carnitas Soft Tacos                       37673        40
Chicken Bowl                             713926       761
Chicken Burrito                          497303       591
Chicken Crispy Tacos                      47382        50
Chicken Salad                             10465         9
Chicken Salad Bowl                       117104       123
Chicken Soft Tacos                        98395       120
Chips                                    208004       230
Chips and Fresh Tomato Salsa             100419       130
Chips and Guacamole                      449959       506
Chips and Mild Fresh Tomato Salsa           279         1
Chips and Roasted Chili Corn Salsa        14999        23
Chips and Roasted Chili-Corn Salsa        17462        18
Chips and Tomatillo Green Chili Salsa     41170        45
Chips and Tomatillo Red Chili Salsa       45556        50
Chips and Tomatillo-Green Chili Salsa     21826        33
Chips and Tomatillo-Red Chili Salsa       18344        25
Crispy Tacos                                434         2
Izze                                      10587        20
Nantucket Nectar                          21814        29
Salad                                      1150         2
Side of Chips                             84769       110
Steak Bowl                               193752       221
Steak Burrito                            328437       386
Steak Crispy Tacos                        35818        36
Steak Salad                                2362         4
Steak Salad Bowl                          29590        31
Steak Soft Tacos                          50177        56
Veggie Bowl                               72270        87
Veggie Burrito                            80962        97
Veggie Crispy Tacos                         668         1
Veggie Salad                               7265         6
Veggie Salad Bowl                         17303        18
Veggie Soft Tacos                          5520         8

If we don't give more specific instruction i.e. indicate _which_ field we want summed up per group, then the default behavior is to _sum up any numeric fields._ So far, these are `order_id` (whose sum isn't really meaningful here) and `quantity` (whose sum does have significance.) We prefer to add up `quantity`'s, so let's do that:

In [73]:
g_sum_qty = g['quantity'].sum()
g_sum_qty.sort_values(ascending=False).head(1)

item_name
Chicken Bowl    761
Name: quantity, dtype: int64

### Step 11. What was the most ordered item in the choice_description column?

In [24]:
chipo.choice_description

0                                                     NaN
1                                            [Clementine]
2                                                 [Apple]
3                                                     NaN
4       [Tomatillo-Red Chili Salsa (Hot), [Black Beans...
                              ...                        
4617    [Fresh Tomato Salsa, [Rice, Black Beans, Sour ...
4618    [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...
4619    [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...
4620    [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...
4621    [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...
Name: choice_description, Length: 4622, dtype: object

Observe that these are either `NaN`'s or else lists (possibly containing nested list(s)). _e.g._

In [38]:
# 'NaN'
print(chipo.choice_description[0])
print(type(chipo.choice_description[0]))

# a simple list
print(chipo.choice_description[2])
print(type(chipo.choice_description[2]))

# a nested list
print(chipo.choice_description[4])
print(type(chipo.choice_description[4]))

nan
<class 'float'>
[Apple]
<class 'str'>
[Tomatillo-Red Chili Salsa (Hot), [Black Beans, Rice, Cheese, Sour Cream]]
<class 'str'>


In [27]:
# ?
# from ast import literal_eval

In [29]:
# This doesn't work because the elements within the lists
# are not "tokenized" as Python strings: instead, they don't
# exist within quotation marks, so 'literal_eval' attempts 
# to interpret them like variables or something else
# choice_desc_list_4 = literal_eval(chipo.choice_description[4])
# choice_desc_list_4

In [50]:
# Instead of working backwards from the column value
# (as the tokenization of a Python list) to the exact
# list which it represents, the following proceeds
# directly to the flattened list containing all 
# elements of the putative list (less principled, but 
# gets where I'm going faster)
import re

def list_token_to_items(list_token):
    if isinstance(list_token, str):
        split_on_seps = re.split(r'[\[,\]]', list_token)
        # trim whitespace
        item_tokens = map(lambda s: s.strip(), split_on_seps)
        # remove empty-string tokens
        item_tokens = filter(lambda s: s != '', item_tokens)
        return list(item_tokens)
    else:
        return []

# e.g. 
choice_desc_str_0 = chipo.choice_description[0]
print(choice_desc_str_0)
choice_desc_items_0 = list_token_to_items(choice_desc_str_0)
print(choice_desc_items_0)

choice_desc_str_2 = chipo.choice_description[2]
print(choice_desc_str_2)
choice_desc_items_2 = list_token_to_items(choice_desc_str_2)
print(choice_desc_items_2)

choice_desc_str_4 = chipo.choice_description[4]
print(choice_desc_str_4)
choice_desc_items_4 = list_token_to_items(choice_desc_str_4)
print(choice_desc_items_4)

nan
[]
[Apple]
['Apple']
[Tomatillo-Red Chili Salsa (Hot), [Black Beans, Rice, Cheese, Sour Cream]]
['Tomatillo-Red Chili Salsa (Hot)', 'Black Beans', 'Rice', 'Cheese', 'Sour Cream']


In [51]:
# transform the 'choice_description' column using our function
# ?
chipo['choice_desc_list'] = chipo['choice_description'].apply(list_token_to_items)
chipo['choice_desc_list']

0                                                      []
1                                            [Clementine]
2                                                 [Apple]
3                                                      []
4       [Tomatillo-Red Chili Salsa (Hot), Black Beans,...
                              ...                        
4617    [Fresh Tomato Salsa, Rice, Black Beans, Sour C...
4618    [Fresh Tomato Salsa, Rice, Sour Cream, Cheese,...
4619    [Fresh Tomato Salsa, Fajita Vegetables, Pinto ...
4620     [Fresh Tomato Salsa, Fajita Vegetables, Lettuce]
4621    [Fresh Tomato Salsa, Fajita Vegetables, Pinto ...
Name: choice_desc_list, Length: 4622, dtype: object

In [53]:
# can I make pandas Series out of these 'object'-dtype'd values?
chipo_desc_list_4 = pd.Series(
    np.asarray(chipo.choice_desc_list[4]))
chipo_desc_list_4

0    Tomatillo-Red Chili Salsa (Hot)
1                        Black Beans
2                               Rice
3                             Cheese
4                         Sour Cream
dtype: object

In [55]:
chipo_all_items = pd.concat(
    [pd.Series(np.asarray(chipo.choice_desc_list[j]))
     for j in range(len(chipo))])

In [56]:
chipo_all_items.value_counts()

Rice                                    2389
Cheese                                  2281
Lettuce                                 1742
Sour Cream                              1711
Black Beans                             1342
Fresh Tomato Salsa                      1046
Guacamole                               1037
Fajita Vegetables                        722
Pinto Beans                              582
Roasted Chili Corn Salsa                 457
Fresh Tomato Salsa (Mild)                351
Tomatillo Red Chili Salsa                325
Fajita Veggies                           302
Roasted Chili Corn Salsa (Medium)        270
Tomatillo-Red Chili Salsa (Hot)          259
Tomatillo Green Chili Salsa              230
Diet Coke                                134
Tomatillo-Green Chili Salsa (Medium)     128
Coke                                     123
Sprite                                    77
Lemonade                                  33
Fresh Tomato (Mild)                       31
Coca Cola 

The most ordered item in the `choice_description` column _if we count within all the lists which each value represents_ is 'Rice', followed by 'Cheese'.

Otherwise (i.e. if we just care about the string values themselves), we would just need the following. _This is the actual solution that was expected for this exercise:_

In [80]:
choice_desc = chipo.groupby('choice_description').sum()
choice_desc.sort_values(by='quantity', ascending=False)['quantity'].head(5)

choice_description
[Diet Coke]                                                               159
[Coke]                                                                    143
[Sprite]                                                                   89
[Fresh Tomato Salsa, [Rice, Black Beans, Cheese, Sour Cream, Lettuce]]     49
[Fresh Tomato Salsa, [Rice, Black Beans, Cheese, Sour Cream]]              42
Name: quantity, dtype: int64

which shows that 'Diet Coke' was most common.

### Step 12. How many items were orderd in total?

In [81]:
item_ordered_total = chipo.quantity.sum()
item_ordered_total

4972

### Step 13. Turn the item price into a float

#### Step 13.a. Check the item price type

In [83]:
chipo.dtypes

order_id               int64
quantity               int64
item_name             object
choice_description    object
item_price            object
choice_desc_list      object
dtype: object

#### Step 13.b. Create a lambda function and change the type of item price

In [86]:
# discard the leading currency symbol
item_price_to_float = lambda item_price: float(item_price[1:])
chipo['item_price'] = chipo.item_price.apply(item_price_to_float)

#### Step 13.c. Check the item price type

In [87]:
chipo.dtypes

order_id                int64
quantity                int64
item_name              object
choice_description     object
item_price            float64
choice_desc_list       object
dtype: object

_`item_price` is now a `float64`!_

### Step 14. How much was the revenue for the period in the dataset?

In [88]:
# add up all `item_price` values
tot_revenue = chipo.item_price.sum()
tot_revenue

34500.16

$34,500.16

_I think the "official" published solution is **incorrect** here because it is multiplying `item_price` by `quantity` for each row. But when comparing different rows that have the same `item_name`, it appears that the `item_price` already factors in the `quantity` purchased in the order. e.g.:_

In [110]:
chipo[(chipo['order_id'] >= 105)  & (chipo['order_id'] <= 106)
                                  & (chipo['item_name'] == 'Chicken Burrito')
    ].loc[:, ['quantity', 'item_name', 'item_price']]

quantity        item_name  item_price
241         2  Chicken Burrito       17.50
242         1  Chicken Burrito        8.75

### Step 15. How many orders were made in the period?

In [90]:
n_orders = chipo.order_id.nunique()
n_orders

1834

### Step 16. What is the average revenue amount per order?

In [91]:
# Solution 1
# Divide `tot_revenue` just computed by `n_orders` just computed
tot_revenue / n_orders
# ~ $18.81

18.811428571428575

In [95]:
# Solution 2
# Group `chipo` dataframe by `order_id`, get the series which 
# reports total price of each `order_id`, then get average of this
# series
chipo_orders = chipo.groupby('order_id')
chipo_order_prices = chipo_orders.item_price.sum()

In [97]:
type(chipo_order_prices)

pandas.core.series.Series

In [98]:
# e.g. 
chipo_order_prices.head(5)

order_id
1    11.56
2    16.98
3    12.67
4    21.00
5    13.70
Name: item_price, dtype: float64

In [100]:
# Now:
avg_rev_per_order = chipo_order_prices.mean()
avg_rev_per_order
# Also ~ $18.81!

18.81142857142869

### Step 17. How many different items are sold?

In [101]:
n_unique_item_names = chipo.item_name.nunique()
n_unique_item_names

50

_50 distinct items, going by `item_name`, are sold._